From f1da5efbf1f96f3648d0e8a7daa79aecf9261e1b Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Wed, 5 Aug 2015 23:36:39 +0100
Subject: [PATCH 15/15] Force linking to libgcc_s_dw2-1.dll [deprecated]

.. by adding to specs:
*libgcc
--undefined=___deregister_frame_info
--undefined=___register_frame_info

These flags get added twice to the ld command line
because of %G appearing twice in:
*link_gcc_c_sequence
%G %L %G
so this should be tidied up really.

20150805 update:

Looking at this again, it seems that the
SHARED_LIBGCC_UNDEFS_SPEC stuff added in commit 7d804f
does the same thing, so my hack may not be needed. It
would appear that for some time, DWARF2_UNWIND_INFO
was not defined, but it's OK now. Latest specs have:
*shared_libgcc_undefs:
%{shared-libgcc: -u ___register_frame_info -u ___deregister_frame_info}
---
 gcc/config/i386/cygwin.h  | 6 +++++-
 gcc/config/i386/mingw32.h | 8 ++++++--
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/gcc/config/i386/cygwin.h b/gcc/config/i386/cygwin.h
index 2186937..7f71ccc 100644
--- a/gcc/config/i386/cygwin.h
+++ b/gcc/config/i386/cygwin.h
@@ -55,6 +55,10 @@ along with GCC; see the file COPYING3.  If not see
     fvtable-verify=std:vtv_end.o%s} \
    crtend.o%s"
 
+/* There is a bug when building i686 dw-2 exceptions
+   where gcc_s gets stripped which this works around */
+#define PREVENT_STRIP_REG_FRAME_INFO "--undefined=___deregister_frame_info --undefined=___register_frame_info"
+
 /* Normally, -lgcc is not needed since everything in it is in the DLL, but we
    want to allow things to be added to it when installing new versions of
    GCC without making a new CYGWIN.DLL, so we leave it.  Profiling is handled
@@ -65,7 +69,7 @@ along with GCC; see the file COPYING3.  If not see
  %{static|static-libgcc:-lgcc -lgcc_eh} \
  %{!static: \
    %{!static-libgcc: \
-     -lgcc_s -lgcc \
+     -lgcc_s " PREVENT_STRIP_REG_FRAME_INFO " -lgcc \
     } \
   } "
 #else
diff --git a/gcc/config/i386/mingw32.h b/gcc/config/i386/mingw32.h
index 0950304..ca73137 100644
--- a/gcc/config/i386/mingw32.h
+++ b/gcc/config/i386/mingw32.h
@@ -120,6 +120,10 @@ along with GCC; see the file COPYING3.  If not see
   %{shared|mdll: " SUB_LINK_ENTRY " --enable-auto-image-base} \
   %(shared_libgcc_undefs)"
 
+/* There is a bug when building i686 dw-2 exceptions
+   where gcc_s gets stripped which this works around */
+#define PREVENT_STRIP_REG_FRAME_INFO "--undefined=___deregister_frame_info --undefined=___register_frame_info"
+
 /* Include in the mingw32 libraries with libgcc */
 #ifdef ENABLE_SHARED_LIBGCC
 #define SHARED_LIBGCC_SPEC " \
@@ -128,9 +132,9 @@ along with GCC; see the file COPYING3.  If not see
    %{!static-libgcc: \
      %{!shared: \
        %{!shared-libgcc:-lgcc -lgcc_eh} \
-       %{shared-libgcc:-lgcc_s -lgcc} \
+       %{shared-libgcc:-lgcc_s " PREVENT_STRIP_REG_FRAME_INFO " -lgcc} \
       } \
-     %{shared:-lgcc_s -lgcc} \
+     %{shared:-lgcc_s " PREVENT_STRIP_REG_FRAME_INFO " -lgcc} \
     } \
   } "
 #else
-- 
2.8.1

